velox/exec/tests/utils/TestValue.h

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <gtest/gtest.h>
+#include "velox/common/caching/SsdCache.h"
+#include "velox/core/Expressions.h"
+#include "velox/core/PlanNode.h"
+#include "velox/exec/tests/utils/QueryAssertions.h"
+#include "velox/type/Variant.h"
+#include "velox/vector/FlatVector.h"
+#include "velox/vector/tests/VectorMaker.h"
+#include "velox/vector/tests/VectorTestBase.h"
+
+namespace facebook::velox::exec::test {
+
+/// The test facility used to inject callback hook in production execution point
+/// to get, change and verify various internal execution state for unit test.
+/// The hook injected in production code path will be opt out in release build
+/// and only be executed in debug build. Please follow the example in
+/// TestValueTest.cpp for use.
+class TestValue {
+ public:
+  /// Enable the test value injection.
+  static void enable();
+  /// Disable the test value injection.
+  static void disable();
+  /// Check if the test value injection is enabled or not.
+  static bool enabled();
+
+  using InjectionCallback = std::function<void(void*)>;
+
+  /// Invoked by the test code to register a callback hook at the specified
+  /// execution point. 'injectionPoint' is a string to identify the the
+  /// execution point which could be formed by concatenating namespace, class
+  /// name, method name plus optional actual action within a method.
+  /// 'injectionCb' is the injected callback hook.
+  static void set(
+      const std::string& injectionPoint,
+      InjectionCallback injectionCb);
+  /// Invoked by the test code to unregister a callback hook at the specified
+  /// execution point.
+  static void clear(const std::string& injectionPoint);
+  /// Invoked by the production code to try to invoke the test callback hook
+  /// with 'testData' if there is one registered at the specified execution
+  /// point. 'testData' can capture the production execution state.
+  static void adjust(
+      const std::string& injectionPoint,
+      void* testData);
```

Comment:
`const void*`

Reply:
We might want to modify the production state in a test injection callback.